### PR TITLE
Comparing dmlc/treelite#357 and dmlc/treelite#359 

### PR DESCRIPTION
**Pull request by hcho3**

### CODE REVIEW

File: src/gtil/predict.cc [hunk 1]
```diff
@@ -76,93 +76,109 @@ inline std::size_t PredictImplInner(const treelite::ModelImpl<ThresholdType, Lea
...
-  // This is guaranteed to be at most GetPredictOutputSize(num_row=1).
   std::size_t output_size_per_row;
   if (pred_transform) {
-    std::vector<float> temp(treelite::gtil::GetPredictOutputSize(&model, 1));
+    std::vector<float> temp_before_transform(num_class);
+    std::vector<float> temp_after_transform(num_class);
     PredTransformFuncType pred_transform_func
         = treelite::gtil::LookupPredTransform(model.param.pred_transform);
-    output_size_per_row = pred_transform_func(model, sum_tloc.data(), temp.data());
+    output_size_per_row = pred_transform_func(model, temp_before_transform.data(),
```
Comment (wphicks) [inline src/gtil/predict.cc:96]: Is there any reason we can't do this transformation in-place?

File: src/gtil/predict.cc [hunk 2]
```diff
@@ -76,93 +76,109 @@ inline std::size_t PredictImplInner(const treelite::ModelImpl<ThresholdType, Lea
...
-      float average_factor;
-      if (model.task_type == treelite::TaskType::kMultiClfGrovePerClass) {
-        TREELITE_CHECK(task_param.grove_per_class);
-        TREELITE_CHECK_EQ(task_param.leaf_vector_size, 1);
-        TREELITE_CHECK_GT(task_param.num_class, 1);
-        TREELITE_CHECK_EQ(num_tree % task_param.num_class, 0)
-          << "Expected the number of trees to be divisible by the number of classes";
-        int num_boosting_round = num_tree / static_cast<int>(task_param.num_class);
-        average_factor = static_cast<float>(num_boosting_round);
+  treelite::threading_utils::ParallelFor2D(std::size_t(0), num_row, std::size_t(0), num_tree,
```
Comment (wphicks) [inline src/gtil/predict.cc:103]: Really like this construct!

**Review by wphicks — APPROVED** [review]
Really nice work! Had just one question, which I think is just me missing something. Good to merge in my book, so long as the perf benchmarks look good.

File: src/gtil/predict.cc (hunk 1 above)
Comment (chyunsu3) [reply src/gtil/predict.cc:96]: The input and the output have potentially different size. The transformer can be a reduction operator.

**Comment by chyunsu3** [comment]
I updated #359, now that https://github.com/dmlc/treelite/pull/357 is part of the `mainline` branch.

**Pull request closed without merging**